### ISSUE TITLE
last_work.ipynb

### ISSUE THREAD

**Issue by pingpong-shige** [pingpong-shige/Python_practice#1]
# 各モジュールインポート
import numpy as np
import pandas as pd
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler, Imputer
from sklearn.feature_selection import RFECV
from sklearn.decomposition import PCA
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier,GradientBoostingClassifier
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.externals import joblib

# 各パラメータセット
# file_model = 'dm_for_model'
# file_score = 'dm_for_fwd'
# ohe_cols = ['mode_category']
# id_col = 0
# y_col = -1
file_model = 'final_hr_analysis_train'
file_score = 'final_hr_analysis_test'
ohe_cols = ['sales', 'salary']
id_col = 0
y_col = 1
id = 'aijc1204'

# モデリング用データ読込
df = pd.read_csv('./data/' + file_model + '.csv', header = 0)
ID = df.iloc[:, [id_col]]
y = df.iloc[:, [y_col]]
X = df.iloc[:, 2:]
# print('data X:', X.shape)
# print('data y:', y.shape)

# One-hot エンコーディング
X_ohe = pd.get_dummies(X, dummy_na = True, columns = ohe_cols)
X_ohe = X_ohe.dropna(axis = 1, how = 'all')
X_ohe_columns = X_ohe.columns.values
# print('one-hot encoding:', X_ohe_columns.shape)

# 欠損値補完
imp = Imputer(missing_values = 'NaN', strategy = 'mean', axis = 0)
imp.fit(X_ohe)
X_ohe = pd.DataFrame(imp.transform(X_ohe), columns = X_ohe_columns)
# print('null imputation:', X_ohe.shape)

# 特徴選択
selector = RFECV(estimator=RandomForestClassifier(random_state = 0), step = 0.05)
selector.fit(X_ohe, y)
X_ohe_selected = selector.transform(X_ohe)
X_ohe_selected = pd.DataFrame(X_ohe_selected, columns=X_ohe_columns[selector.support_])
# print('feature selection:', X_ohe_selected.shape)
X_ohe_selected.head()

# 交叉検証
X_train, X_test, y_train, y_test = train_test_split(X_ohe_selected, y, test_size=0.2, random_state=1)
# shaping
y_train = y_train.as_matrix().ravel()
y_test = y_test.as_matrix().ravel()

# アルゴリズム候補
pipelines = {
    'kn': Pipeline([('scl', StandardScaler()), ('est', KNeighborsClassifier())]),
    'knp': Pipeline([('scl', StandardScaler()), ('pca', PCA(n_components=10, random_state=1)), ('est', KNeighborsClassifier())]),
    'lr': Pipeline([('scl', StandardScaler()), ('est', LogisticRegression(random_state=1))]),
    'lrp': Pipeline([('scl', StandardScaler()), ('pca', PCA(n_components=10, random_state=1)), ('est', LogisticRegression(random_state=1))]),
    'rf': Pipeline([('scl', StandardScaler()), ('est', RandomForestClassifier(random_state=1))]),
    'rfp': Pipeline([('scl', StandardScaler()), ('pca', PCA(n_components=10, random_state=1)), ('est', RandomForestClassifier(random_state=1))]),
    'gb': Pipeline([('scl', StandardScaler()), ('est', GradientBoostingClassifier(random_state=1))]),
    'gbp': Pipeline([('scl', StandardScaler()), ('pca', PCA(n_components=10, random_state=1)), ('est', GradientBoostingClassifier(random_state=1))])
}

# モデリング検証
models={}
for pipe_name, pipeline in pipelines.items():
        pipeline.fit(X_train, y_train)
        results = cross_val_score(pipeline, X_train, y_train, scoring='roc_auc', cv=5)
#         print(pipe_name + ' cv score:', np.average(results), '+-', np.std(results))
        models[pipe_name]=np.average(results)
print('-- Model Ranking ----------------------------------------')
# print(models)
models=sorted(models.items(), reverse=True, key=lambda x:x[1])
# print(models)
for model in models:
    print(model[0] + ' cv score:', model[1])

# モデル選択
model_name = models[0][0]
print('-- Best Model ---------------------------------------------')
print('selected model:', model_name)
clf = pipelines[model_name]
# print(clf)

if len(file_score)>0:
    # スコア用データ読込
    dfs = pd.read_csv('./data/' + file_score + '.csv', header=0)
    IDs = dfs.iloc[:,[id_col]] 
    Xs = dfs.iloc[:, 2:]
    Xs_ohe = pd.get_dummies(Xs, dummy_na = True, columns = ohe_cols)

    # ヘッダだけのDataFrame作成
    cols_m = pd.DataFrame(None, columns = X_ohe_columns, dtype = float)
    Xs_exp = pd.concat([cols_m, Xs_ohe])

    # モデリング時にのみ出てきた列を追加
    Xs_exp.loc[:, list(set(X_ohe_columns) - set(Xs_ohe.columns.values))] = \
      Xs_exp.loc[:, list(set(X_ohe_columns) - set(Xs_ohe.columns.values))].fillna(0, axis = 1)

    # スコアリング時にのみ出てきた列を削除
    Xs_exp = Xs_exp.drop(list(set(Xs_ohe.columns.values) - set(X_ohe_columns)), axis = 1)

    # 並び順を揃える
    Xs_exp = Xs_exp.reindex_axis(X_ohe_columns, axis = 1)
    Xs_exp = pd.DataFrame(imp.transform(Xs_exp), columns = X_ohe_columns)
    Xs_exp_selected = Xs_exp.loc[:, X_ohe_columns[selector.support_]]

# モデリング
clf.fit(X_ohe_selected, y)
joblib.dump(clf, './model/'+ model_name + '.pkl')
# results = cross_val_score(clf, X_ohe_selected, y.as_matrix().ravel(), scoring='roc_auc', cv=5)
results = cross_val_score(clf, X_ohe_selected, y, scoring = 'roc_auc', cv = 5)
print('-- Last Score ---------------------------------------------')
print('cv score:', np.average(results), '+-', np.std(results))

# スコアリング
if len(file_score)>0:
    score = pd.DataFrame(clf.predict_proba(Xs_exp_selected)[:, 1], columns = ['left'])
    IDs.join(score).to_csv('./data/'+  id + '.csv', index=False)

# # model profile
# imp = pd.DataFrame([clf.named_steps['est'].feature_importances_], columns=X_ohe_columns[selector.support_])
# imp.T.to_csv('./data/'+  model_name + '_feature_importances.csv', index=True)
